fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/InheritedPermissionEvaluator.java

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright 2017 The University of North Carolina at Chapel Hill
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.unc.lib.dl.acl.fcrepo4;
+
+import static edu.unc.lib.dl.acl.util.PrincipalClassifier.classifyPrincipals;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import edu.unc.lib.dl.acl.util.Permission;
+import edu.unc.lib.dl.fedora.ContentPathFactory;
+import edu.unc.lib.dl.fedora.PID;
+
+/**
+ * Evaluates the permissions of agents on specific objects, taking into account
+ * inherited access control information
+ * 
+ * @author bbpennel
+ *
+ */
+public class InheritedPermissionEvaluator {
```

Comment:
I agree that it would be easier to read/modify if the staff and patron parts were processed separately, and I initially went back and forth on doing it that way.

I ended up implementing it as one uber loop so as to minimize the number of iterations/lookups that would be necessary under the assumption that this method was going to be called on every request (and every object in the request).  So my thinking was to prioritize efficiency over readability a bit in this case.  However, I haven't actually measured how significant a difference there is between the approaches, which is probably not huge since the query results are cached.  So it may not be worth the added complexity.

Principal classification is actually in a separate utility though, its just being referenced statically here.